### ISSUE TITLE
off by one problem when generating pyramidal tiff

### ISSUE THREAD

**Issue by beaudet** [libvips/libvips#659]
Original issue noted here: https://github.com/ruven/iipsrv/issues/98 but copied below since this is really a vips issue, not an IIP issue.

I have confirmed that there is an ongoing resize bug in vips that is reproducible on the latest up-to-date RHEL6 and RHEL7 systems using either vips 7.x or the latest 8.4 version of vips. See the differences between the images below which were taken from the pyramidal tiff. The first was created using imagemagick and the second with vips 8.4.5 on RHEL7. Aside from being slower, we had found a few years back that IM creates intermittent corrupted ptifs that IIP can't open. I think the only option is to find the bug in vips or the library it's using and fix it - or switch to IM if the PTIF generation has been fixed since we last looked at it about 4 or 5 years ago. Any advice appreciated.

One of the resolutions from an ImageMagick generated pyramidal tiff image
![image](https://cloud.githubusercontent.com/assets/77699/26126262/f01b4f70-3a52-11e7-925e-acbe82391fd2.png)

and same resolution from vips generated pyramidal tiff (note flatness on right side and bottom) image
![image](https://cloud.githubusercontent.com/assets/77699/26126266/f58f564a-3a52-11e7-9d49-418358a99ffd.png)

command used to generate the ptif: 
vips im_vips2tiff ./source_tiff_file.tif vips_regen_8.4.5_rhel7.ptif:jpeg:90,tile:256x256,pyramid

info about the original TIFF source: TIFF 11038x11038 11038x11038+0+0 8-bit DirectClass 365.5MB

command used to generate the output above

convert vips_regen_8.4.5_rhel7.ptif -quality 100 vips_regen_8.4.5_rhel7.jpg

This produces one jpg for each of the 6 or 7 resolutions that vips generates in the ptif.  All of the exported jpegs from a ptif generated by vips exhibit this problem whereas none of the exported jpgs generated with imagemagick do.

**Comment by beaudet** [comment]
Here's the configure info:

* build options
native win32:                           no
native OS X:                            no
open files in binary mode:              no
enable debug:                           no
build deprecated components:            yes
build docs with gtkdoc:                 no
gobject introspection:                  yes
build vips7 Python binding:             no
install vips8 Python overrides:         no
  (requires pygobject-3.12.0 or later)
build radiance support:                 yes
build analyze support:                  yes
build PPM support:                      yes

* optional dependencies
use fftw3 for FFT:                      no
Magick package:
file import with libMagick:             no
accelerate loops with orc:              no
  (requires orc-0.4.11 or later)
ICC profile support with lcms:          no
file import with OpenEXR:               no
file import with OpenSlide:             no
  (requires openslide-3.3.0 or later)
file import with matio:                 no
PDF import with poppler-glib:           no
  (requires poppler-glib 0.16.0 or later)
SVG import with librsvg-2.0:            no
  (requires librsvg-2.0 2.34.0 or later)
zlib:                                   yes
file import with cfitsio:               no
file import/export with libwebp:        no
  (requires libwebp-0.1.3 or later)
text rendering with pangoft2:           no
file import/export with libpng:         no
  (requires libpng-1.2.9 or later)
file import/export with libtiff:        yes (pkg-config libtiff-4)
file import/export with giflib:         no
file import/export with libjpeg:        no
image pyramid export:                   no
  (requires libgsf-1 1.14.26 or later)
use libexif to load/save JPEG metadata: no

Vips-8.0.typelib will install to /usr/local/lib/girepository-1.0, but your
system repository seems to be /usr/lib64/girepository-1.0.
You may need to add this directory to your typelib path, for example:

        export GI_TYPELIB_PATH="/usr/local/lib/girepository-1.0"

**Comment by beaudet** [comment]
This seems to be a problem with the way vips is sampling from the next higher resolution.  Since it literally discards every other column of pixels and then averages the pixel with the pixel from one scanline below, this is causing the image to shift to both the right and down.  The effect is compounded from one resolution to the next since sampling works in succession rather than from the original full size source.  I think the general nearest neighbor style approach would be fine with tweaks to how the resampling is performed but as written, it's wrong.  I'm referring to this definition in libvips/iofuncs/region.c:

#define SHRINK_TYPE_INT( TYPE ) \
        for( x = 0; x < target->width; x++ ) { \
                TYPE *tp = (TYPE *) p; \
                TYPE *tp1 = (TYPE *) (p + ls); \ <-- I believe that only averaging with the the pixel one scanline below results in cropping the image at the bottom
                TYPE *tq = (TYPE *) q; \
                \
                for( z = 0; z < nb; z++ ) { \
                        int tot = tp[z] + tp[z + nb] +  \
                                tp1[z] + tp1[z + nb]; \
                        \
                        tq[z] = tot >> 2; \
                } \
                \
                /* Move on two pels in input. \  <---- this discards data and results in chopping off the last column of image data rather than averaging it - this has the effect of cropping the image on the right side
                 */ \
                p += ps << 1; \
                q += ps; \
        }

**Comment by jcupitt** [comment]
Hi @beaudet, 

I think my instinct would be to blame the strict round down tiffsave is doing when it makes the layer below:

https://github.com/jcupitt/libvips/blob/master/libvips/foreign/vips2tiff.c#L393

On average you'll lose 0.5 pixels per layer off the right and bottom, so in 10 layers 5 pixels will go. Would that be enough to explain your missing pixels?

I think the simplest solution might be to use `rint()`, which rounds up and down alternately, and so should preserve size. Or maybe calculate each layer size from the original image size, rather than clipping repeatedly.

The advantage of the strict round-down is that it avoids having to handle half pixels at the edges. Some code would need to be added to cope with this. The deepzoom writer has something that could be nicked.

https://github.com/jcupitt/libvips/blob/master/libvips/foreign/dzsave.c#L1311

I'll see if I can copy something like that over.

On `SHRINK_TYPE_INT`, it's not discarding pixels, is it? Each pixel in a layer is the average of the corresponding four pixels in the layer above.

https://github.com/jcupitt/libvips/blob/master/libvips/iofuncs/region.c#L1170

That's a standard x2 box downsize, I don't think there's a problem there. 

The `tot >> 2` should probably be `(tot + 1) >> 2`.

**Comment by beaudet** [comment]
John, thanks for your quick response. The algorithm seems perfectly fine when applied once. When I look at all the resolutions, no stretching is apparent at 50% of the original (at least that I can detect), but the effect begins to show up at 25% and gets worse from there so it's likely to be associated with the operation being repeated many times. I think you're right that it's  a rounding issue and probably one associated with the averaging of pixel values. I think using the true floating point center of the source vs target to weight the pixels would work but that might be a little slower and would probably no longer be a pure implementation of the box downsize filter. 

I missed the section in the code that's actually averaging pixels along the X dimension but will take another look.

**Comment by beaudet** [comment]
Also, on an unrelated topic, do you find that performing pointer addition rather than using indexes into an array performs better and if so, by approximately how much? Similar question regarding the choice to use a define rather than casts in the SHRINK algorithm. Just curious.

**Comment by jcupitt** [comment]
The pointer addition is an optimisation called strength reduction. You could write:

```C
for (i = 0; i < 1000; i++)
    for (j = 0; j < 1000; j++)
        sum += array[i + j];
```

or the equivalent:

```C
for (i = 0; i < 1000; i++) {
    restrict int *p = array + i;

    for (j = 0; j < 1000; j++)
        sum += p[j];
}
```

Now you add `i` once rather than 1000 times. Compilers will do some optimisations like this automatically, but when loops get complex they often fail. gcc will also automatically vectorise version 2, which is nice. It might vectorise v1, but I'd be a little surprised. 

The defines serve the same purpose: help the compiler vectorise by giving it the simplest code path you can. 

There's no translation or stretching, just truncation along the right and bottom edge of the image. The averaging will tend to make the image slightly darker, since there's a `+1` missing from before the shift, but it will not displace the image.

**Comment by jcupitt** [comment]
Oh heh, of course it won't vectorise the inner shrink loop, it's too small. Anyway, you need to instantiate many version of the code to handle the different pixel types, and a macro is the simplest way to do taht.

**Comment by beaudet** [comment]
In resolution 1 (50% of original size), ImageMagick and vips are pixel-aligned and there seems to be no stretching or removal of any columns.  These jpeg levels might not be the same so don't try to compare for quality here - I'm just looking at the pixel alignment.

Vips generated and zoomed in 500% at right edge of resolution 1
![image](https://cloud.githubusercontent.com/assets/77699/26206443/db5b2072-3bb2-11e7-9087-438611da7a82.png)

ImageMagick generated and zoomed in at 500% at right edge of resolution 1
![image](https://cloud.githubusercontent.com/assets/77699/26206432/d1563ff8-3bb2-11e7-989f-1028e9e39920.png)

At level 2, the differences start to emerge - if you open these into two tabs, you can see that the entire image is being stretched to the right and down in the vips version.

vips at level 2:
![image](https://cloud.githubusercontent.com/assets/77699/26206685/7c32e8fe-3bb3-11e7-808a-479751118bca.png)

image magick at level 2:
![image](https://cloud.githubusercontent.com/assets/77699/26206675/77752d22-3bb3-11e7-83da-fe756646e63d.png)

**Comment by beaudet** [comment]
Here's a link to the 300+MB original bits in case you want to try to replicate the issue:

https://media.nga.gov/public/objects/4/1/5/8/1/tests/divs_original/A15324_XBD.tif

**Comment by beaudet** [comment]
Thanks John.

> There's no translation or stretching, just truncation along the right and bottom edge of the image. The averaging will tend to make the image slightly darker, since there's a +1 missing from before the shift, but it will not displace the image.

I see that int tot = tp[z] + tp[z + nb] + tp1[z] + tp1[z + nb]; is taking the pixel to the right as you say.  So, I guess the question is why 

The current algorithm sets target pixel Q0 = average of source pixels P0+P1+P2+P3
---|----
P0 | P1|
---|--- |   =>   Q0 
P2 | P3|
--------

That seems perfectly correct to me.

Like you already said, I think the issue is the size of the canvas that is calculated from one step to the other.  It alternates from even to odd numbers and loses a half pixel whenever dividing an odd resolution by half.  For example, if the canvas starts as 11038 for resolution 0, then resolution 1 is 5519, but resolution 2 becomes 2759 which, when filling in the last pixel at pixel index 2758 computes with 2*2758 = 5516 whereas it really should be taking it from 5517 at that point.  This is what's leaving out the last column of data.  So, it's losing 0.5 pixels every other layer like you already said. 

I think there are going to be problems with the existing algorithm unless it's changed to a weighted algorithm based on distance from pixel center to computed scale center.  It seems that even if edges are treated specially, you're still going to have drop a column of pixels somewhere in the image without weighting.  And that's going to result in shifting the rest of the image, right?

Maybe an initial downscaling operation could be performed on resolution 0 or on resolution 1 so that it becomes an even multiple of 2^(R-1) where R is the number of resolutions?  That way, the existing resampling algorithm will work perfectly for successive tile size without losing any pixels.  For example, I've re-sampled the original file to be 11008 pixels instead of 11038.  11008 is evenly divisible by 2^(7-1) all the way to 172 which was the originally computed smallest tile size.  It works perfectly with the existing algorithms in place and matches the imagemagick generated ptif in terms of scale and pixel placement.

**Comment by jcupitt** [comment]
The dzsave saver handles this by expanding odd-sized layers. I'm planning to try this in the tiff writer. I'll make a branch.

**Comment by jcupitt** [comment]
Oh, you're right, it should be +2, I'll change it.

**Comment by beaudet** [comment]
Sounds good John, and thanks!  We're all looking forward to finally having an end-to-end solution that can produce high quality derivatives on-the-fly with IIP Image server so we no longer have to create and track web-specific derivatives in our DAM and associated systems.  This vips fix is the last piece of that puzzle.

**Comment by jcupitt** [comment]
Here's a branch that rounds up rather than down:

https://github.com/jcupitt/libvips/tree/tiff2vips-round-layers-up

What do you think? You might find layers can be too large now, perhaps round to nearest would be better. 

Here's what I get for:

```
$ vips tiffsave A15324_XBD.tif x.tif --pyramid --tile
$ vips copy x.tif[page=6] x6.png
```

![x6](https://cloud.githubusercontent.com/assets/580843/26275442/23d04f66-3d58-11e7-9190-f2bc95d74965.png)

**Comment by beaudet** [comment]
I'm a little concerned about the effect that copying the rightmost column will have on the end result and on successive layers, but I'll try with a bunch of different resolutions and see what happens.

**Comment by jcupitt** [comment]
The copying happens in the layer above, so all it's doing is changing the rightmost column from a 2x2 average to a 1x2 average.

**Comment by jcupitt** [comment]
Hello again, did you get a chance to look at this? Should I change it to round-to-nearest?

**Comment by beaudet** [comment]
I was assigned three short term critical projects the day I was last going to look at this.  I'm nearing completion of those and hope to take a look this week or next.

**Comment by beaudet** [comment]
It's looking much much better now.  Maybe done? I think you're probably there, but I have a question.  If you pull the two images below into a browser and flash between them, you'll see that the VIPS generated image is actually now slightly smaller than the imagemagick generated image which was created using the same geometry as the computed tile size.  One could argue that's actually more correct - I'm not sure what vips would be filling into the last column to pad the space otherwise - it's certainly not manufacturing pixel data and I see actual data in that last column so presumably there's a bug in IM w.r.t. image resizing too :)

I'll give this version a pirouette and hopefully transition to using it to recompute all of our 80k images soon. 

VIPS
![image](https://user-images.githubusercontent.com/77699/27148841-bdf82cc4-510f-11e7-9e73-2e0fed7640d9.png)


ImageMagick
![image](https://user-images.githubusercontent.com/77699/27148835-b723bf26-510f-11e7-8bc9-f3366789ca8f.png)

**Comment by jcupitt** [comment]
Would you mind testing round to nearest as well? I'll change it tomorrow.

**Comment by beaudet** [comment]
yup, it appears that IM is significantly reducing the vibrancy of the last column - not entirely clear why that is, but might have to do with the re-sampling algorithm it uses by default.  See below for examples with -quality 100 jpeg compression.

imagemagick conversion from original 4003x4003 source file:
![image](https://user-images.githubusercontent.com/77699/27150058-bd705f66-5113-11e7-8f8d-fb90e13389d2.png)

imagemagick conversion / extraction of the tile from vips generated ptif 
![image](https://user-images.githubusercontent.com/77699/27150069-c3a66fce-5113-11e7-8a68-1368eb3491c1.png)

through IIP image server using PNG output
![image](https://user-images.githubusercontent.com/77699/27150074-c8a47fa2-5113-11e7-9e7a-7e87980e8c3f.png)

through IIP image server using JPG output
![image](https://user-images.githubusercontent.com/77699/27150083-cfaa46ce-5113-11e7-8ab7-c53394b5255c.png)

**Comment by beaudet** [comment]
Sure John, no problem, but how would I go about testing round to nearest?  Is there a branch or command line switch for round to nearest?

**Comment by jcupitt** [comment]
I mean I'll update that branch, hang on.

**Comment by beaudet** [comment]
Acknowledged

**Comment by beaudet** [comment]
This seems to be an IIP protocol related bug - the IIIF viewers render fine with the same IIP server in use.  I'll report it to Ruven.

**Comment by ruven** [comment]
This is using a round to nearest vips? This could be an iipmooviewer rather than a vips issue. I think iipmooviewer is assuming the old rounding behaviour of vips when calculating the sizes and tile layout.

**Comment by beaudet** [comment]
currently, it's rounding up - but we have plans to also test round to nearest (in the queue. - John will provide an update for that soon so I can test). It certainly makes sense that it could be a viewer issue.  Interesting that it only happens in one of 50 regenerated images.

**Comment by ruven** [comment]
It'll only happen when the difference in size between a rounded down or up resolution requires an extra tile in the horizontal direction. This should be quite a rare occurrence, so 1 in 50 sounds reasonable.

**Comment by beaudet** [comment]
I've created an issue for this in the IIPMooviewer project: https://github.com/ruven/iipmooviewer/issues/73

**Comment by jcupitt** [comment]
Hi, I pushed some fixes, could you test again? 

Round to nearest doesn't make any sense (duh) since we only round 0.0 and 0.5, so it's the same as round up.

I found a rare problem with black columns, which I've fixed. 

I did some more checking, and this patch will add one extra row or column in some layers in some circumstances, but that's all it does. It sounds like we need to check with Ruven too.

**Comment by jcupitt** [comment]
Some more notes on this: I was being very dumb when I said errors could accumulate -- they can't. Imagine the width as a binary number. With the old behaviour, we were simply shifting right, so errors obviously can't build up. Or in other words, we are reducing precision by the same amount as our rounding error, so they cancel.

Therefore by doing round up instead of down, we cannot add more than one extra column or row to any layer.

I think your test image was unusual in that it was cropped very tightly, and a roundel will show even a single pixel of extra crop.

**Comment by beaudet** [comment]
I'm pretty sure there was more than one column of data missing in the original example,at least that's how it appears. Where does this leave us w.r.t. a solution?  Rather than adding columns can the resize operation distribute pixels with floating point precision? Or could that be a command line option at least? I'd also be fine with reducing the size of the largest layer so that it's an even multiple of the tile size.  Incidentally, it seems OSD might also have an issue with these slightly uneven pyramids.

**Comment by beaudet** [comment]
I think I understand your point. Adding one row isn't going to add rows cumulatively with each layer.  I guess the question is what impact this has on viewers. Are clients following a quasi (or real?) standard for PTIF that mandates "exactly" half? If so, altering that logic is going to be a breaking change. Not necessarily bad if the underlying assumption is bad but I can imagine Javascript authors running into the same problems with resize that libvips is but having far fewer tools to deal with it, so it might be really important to keep it at exactly 1/2.

**Comment by jcupitt** [comment]
I think we should:

* Check with Ruven about IIP and rounding
* Could you test the latest version of the branch? It does behave slightly differently now
* I certainly agree it should not clip along the right and bottom. If we can isolate a case where this happens in the current code, that's something that should be fixed
* If IIP forces us to round down, then I suppose we'll need to shift images left and up by 0.5 pixels to keep any clipping symmetrical ... this will probably make image write quite a bit slower, unfortunately

**Comment by ruven** [comment]
The IIP protocol doesn't really define how rounding should be done. But the protocol was originally designed to support the Flashpix format, whose specification defines how resolutions should be sized: [fpx.pdf: page18, section 2.2.1](http://graphcomp.com/info/specs/livepicture/fpx.pdf), which looks to me like rounding up.

Kakadu also resizes it's JPEG2000 layers with rounding up (or perhaps to nearest) and iipsrv takes this into account to emulate vips TIFF round-down behaviour for clients.

Imagemagick, however, seems to round down, from what I can see.

Internally, iipsrv can handle arbitrary resolution sizes, but many viewers will assume a certain rounding behaviour, including iipmooviewer.

**Comment by beaudet** [comment]
After looking into IIPMooviewer, one of our front-end developers seems to think the problem might be with our latest implementation of IIP server, so I'm guessing that I might have introduced a bug into the server or perhaps a bug exists in the server that only showed up with these rounded up sizes.  I hope to isolate a test case and the first step will be rolling back to the prior version of IIP server to see if that fixes the problem and then go from there.

**Comment by beaudet** [comment]
No big in IIP for this particular behavior.  It's a problem with IIP Mooviewer client as Ruven expected.  I'll document more details in that ticket.

However, back to the originally scheduled program regarding vips behavior in creating PTIFs. 

The more I think about this and understand how the zoom image viewers work, the more I'm convinced that vips should actually have an option to do the heavy lifting of applying a high quality resampling filter when it creates resized images.  Here's my thinking:

The two primary use cases we have for the IIIF image API are to support zoom and to generate arbitrary derivatives on the fly.  IIP will soon have a release that supports Lanczos as well as oversampling which together generate very high quality derivatives.  Those operations happen completely on the server side, although some ambiguity remains when using % specifications with the IIIF API, but let's forget about that for now.  The second use case is zoom clients which typically (for performance reasons) just fetch tiles from the source image processed by vips.  There is no resampling at all performed by image servers in that operation, so these clients are highly reliant on the quality of images being produced by vips.   The alternative would be to force resampling / oversampling instead of just fetching / returning tiles, but that's pretty wasteful when the operation can just be performed once and be done.

So, with respect to what approach to take, I think sticking with the floor() approach vis-a-vis integer arithmetic for the computed size is probably fine, but a Lanczos filter should be applied when downsampling from the original.  I think that's a small price to pay for being able to avoid the same computation over and over again in a zoom server context. 

There also seems to be a lot of inconsistency with how zoom clients compute rounding - some of them round up, some round down and there are protocol dependencies on the server side.  Changing the current round-down behavior to round-up is potentially going to require a lot client fixes.  So, while forcing the resolution ambiguity problem out would probably be a good thing in the long term, it might be a bit too impactful to do unilaterally.  Some of the inconsistency could be fixed by forcing an image to be a multiple of the tile size, but that will necessarily require some resizing for institutions that are trying to use a JP2 as the original and their image source, so resizing to multiple probably isn't a great approach for many - for those that want to do that, it can be scripted easily enough as it is.

So, I'd like to request that vips continue the current approach w.r.t. handling resolutions, but that an optional command line parameter can control the resizing algorithm applied when downsampling from the 2x source page.  Furthermore, I'd like to request that Lanczos 3 be used since it seems to be generally accepted that it's the best all around downsampling filter (and as Ruven has discovered, there's also Intel hardware support for it so it can run very quickly on Intel hardware).   That will ensure the highest possible image quality in PTIFs with no oddities associated with extra columns and rows. 

What do you think John / Ruven?

**Comment by beaudet** [comment]
John, there's also an updated article about rounding between IIIF clients and servers.  It's not specifically about what we're talking about in this article since the info.json call returns all of the resolutions that are available, regardless of how they are computed, but including here as a reference: https://www.jack-reed.com/2016/10/14/rounding-strategies-used-in-iiif.html

**Comment by jcupitt** [comment]
I made a pair of test images:

![a](https://user-images.githubusercontent.com/580843/27381907-81e31826-567c-11e7-8146-5500eb88713b.png)
![b](https://user-images.githubusercontent.com/580843/27381910-87fa4c20-567c-11e7-9590-720eb868a67f.png)

One is a repeated 2x2 box filter (the thing libvips uses at the moment to make pyramid layers), the other is lanczos3 (the thing libvips uses to make image thumbnails). Put them into different tabs and try flipping between them. I won't tell you which is which :) 

The lanczos3 one is slightly better, but it seems a bit marginal to me, and of course there's a huge speed penalty. And it would be quite a bit of work to add a fast x2 lanczos3 downsample.

**Comment by beaudet** [comment]
The second one is clearly better so I assume that's the Lanczos?  I know it's not a trivial amount of work, but I really don't mind using the up-front computational time in exchange for applying Lanczos + oversampling with every request - it's the green thing to do and it will also result in faster zoom server performance.

There's a software implementation of Lanczos available already in the IIP Image code if you want to grab that one for starters.  Ruven is implementing an Intel hardware accelerated version that he reports is 6-8 times faster.  There's a Lanczos pull request in the IIP Image project.  I could take a crack at it as well, but it would probably take me a lot longer.

**Comment by jcupitt** [comment]
libvips has a very fast lanczos3 already, but it's a general-purpose one. This would need one tuned specifically for a x2 downsize. I'm tinkering a bit with this now.

**Comment by ruven** [comment]
I'm not sure I would say it's "clearly" better. The second image is noticeably brighter (average pixel value of 138.56 vs 136.97). But the first seems to be slightly sharper.

> There's a software implementation of Lanczos available already in the IIP Image code 

libvips already has an implementation of Lanczos resizing

**Comment by jcupitt** [comment]
The libvips one, if you're curious:

https://github.com/jcupitt/libvips/blob/master/libvips/resample/reducev.cpp

It does things like run-time code generation and exploits the processor's vector units.

**Comment by beaudet** [comment]
Here's an example of why I say "clearly" better - take a look at the edges of the halos - there are other examples as well.  If we're going to serve up raw tiles in a zoom viewer without performing any resampling operations, I'd prefer to serve them up from the second image.

box filter (exploded with browser)
![image](https://user-images.githubusercontent.com/77699/27384550-3f7ff4b4-565d-11e7-97fd-30fa28f521e9.png)

Lanczos (exploded with browser)
![image](https://user-images.githubusercontent.com/77699/27384558-465e0e2e-565d-11e7-8073-3597331005ac.png)

**Comment by jcupitt** [comment]
Yes, I'm a little puzzled by the brightness difference, I ought to look into it.

**Comment by beaudet** [comment]
thanks John.

**Comment by ruven** [comment]
Encoding time isn't an issue for a lot of people, so maybe we could have a command line parameter or environment variable for vips to switch between using a box filter or lanczos3 for pyramid resizing?

**Comment by jcupitt** [comment]
Here's a branch that adds a x2 lanczos3 shrink:

https://github.com/jcupitt/libvips/tree/add-lanczos3-tiff-pyr

It compiles, but it's untested, and the tiff writer doesn't use it anyway, heh. I'll wire it up.

**Comment by ruven** [comment]
Regarding the rounding down / rounding up issue, iipmooviewer can now get a list of the actual resolutions that exist from iipsrv and so can handle both rounded down and rounded up images. I'll also add a configuration parameter to iipsrv to allow the user to simulate rounding down (or up) if using a client that requires a certain behaviour. This also potentially allows image pyramids to be used which have scaling by factors other than 2.

@jcupitt, for vips, how will you set this up? Have rounding configurable in some way, but with a default?

**Comment by beaudet** [comment]
@jcupitt - John, shall I test the branch above and let you know how it goes or did you want to first wire up the tiff writer?

**Comment by beaudet** [comment]
@jcupitt - ping

**Comment by jcupitt** [comment]
Hi, still tinkering with it, another day or two.

**Comment by jcupitt** [comment]
It seems to be working! I now see:

![x](https://user-images.githubusercontent.com/580843/27761479-859684aa-5e54-11e7-9e3a-f5f6dfebf198.png)

Which is at least clipped symmetrically. That's using lanczos3, so there's a tiny bit more detail than before. It's doing round-down, so it should work with any viewer.

Here's the git master version, for comparison:

![x2](https://user-images.githubusercontent.com/580843/27761516-37606296-5e55-11e7-8e44-90146f5901a4.png)

It's a bit slower, but not horribly. New:

```
$ time vips tiffsave A15324_XBD.tif x.tif --tile --pyramid --compression jpeg
real	0m3.499s
user	0m3.676s
sys	0m0.252s
```

Old:

```
$ time vips tiffsave A15324_XBD.tif x.tif --tile --pyramid --compression jpeg
real	0m2.596s
user	0m2.796s
sys	0m0.224s
```

I'll add a switch for "enable new stuff".

**Comment by beaudet** [comment]
Great news! I should have some time to try this out late next week. Really looking forward to it. Thanks John!

**Comment by ruven** [comment]
Looks good! I presume the rounding used makes no difference to the speed in this case. It would be nice, however, to be able to choose what kind of rounding is done. Maybe you could add an environment variable such as VIPS_ROUNDING?

**Comment by jcupitt** [comment]
We now have two new options: `--fancy` turns on the new lanczos3 downsampling, and `--round-up` makes it round layer sizes up rather than down. 

The names are just for testing, of course, we should think of a better way to control this for the final code.

**Comment by beaudet** [comment]
Thanks so much for this John. This capability  paves the way for us to deliver dynamic images on the fly, i.e. eliminate production of all derivatives save the PTIF.

**Comment by ruven** [comment]
I just tried to use this branch, but it only creates the first 2 pyramid layers irrespective of whether --fancy is enabled.

**Comment by jcupitt** [comment]
Oh dear, maybe I dropped a commit. I'll have a look.

**Comment by jcupitt** [comment]
It's working on the test images I have. What are the pixel dimensions of the image you tried?

**Comment by ruven** [comment]
This is the output of tiffinfo of the image I'm converting from:
```
  Image Width: 15016 Image Length: 11741
  Resolution: 118.103, 118.103 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: RGB color
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 3
  Rows/Strip: 372
  Planar Configuration: single image plane
```

And the result from using tiffsave with options --tile --pyramid --compression deflate  --vips-progress --predictor horizontal  --tile-width 256 --tile-height 256

```
TIFF Directory at offset 0x12141e7c (303308412)
  Image Width: 15016 Image Length: 11741
  Tile Width: 256 Tile Length: 256
  Resolution: 118.103, 118.103 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: AdobeDeflate
  Photometric Interpretation: RGB color
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 3
  Planar Configuration: single image plane
  Predictor: horizontal differencing 2 (0x2)
TIFF Directory at offset 0x17137928 (387152168)
  Subfile Type: reduced-resolution image (1 = 0x1)
  Image Width: 7508 Image Length: 5870
  Tile Width: 256 Tile Length: 256
  Resolution: 118.103, 118.103 pixels/cm
  Bits/Sample: 8
  Compression Scheme: AdobeDeflate
  Photometric Interpretation: RGB color
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 3
  Planar Configuration: single image plane
  Predictor: horizontal differencing 2 (0x2)
```
So, only 2 resolution layers are created. This is using your up to date add-lanczos3-tiff-pyr branch.

**Comment by jcupitt** [comment]
Thanks! I think it's working now, I was missing a clip, could you try again?

**Comment by beaudet** [comment]
It appears there's still a problem with lanczos and the round-up option w.r.t. centering the image.  In addition, with Lanczos3 it appears that the image is smooshed a bit on the top and left now.

--fancy --round-up copying page 6 (173x173)
![image](https://user-images.githubusercontent.com/77699/28018695-47f9bef6-654c-11e7-949a-f31faaede530.png)


-- fancy (no round up) page 6 (172x172)
![image](https://user-images.githubusercontent.com/77699/28018713-5517d672-654c-11e7-8dec-7da2110dc754.png)

image magick resize (default options to same resolution - 172x173) - note the placement of the image within the bounding box - it's centered which matches the original crop
![image](https://user-images.githubusercontent.com/77699/28018935-1c41b04c-654d-11e7-8dbe-2261d9db0ace.png)

**Comment by ruven** [comment]
> Thanks! I think it's working now, I was missing a clip, could you try again?

Thanks, the pyramid is now correctly built.

**Comment by beaudet** [comment]
Hi John, any updates on the (off by one?) issue with round-up above and smooshing of the image against the sides?

**Comment by jcupitt** [comment]
I think round up is probably a bad idea, since it will break at least some clients. 

I've changed it to do yet another variation: this time there's a 0.5 pixel shift down and right, so it's no longer clipping slightly along the top and left edge.

Default (ie. plain 2x2 downsample):
![x](https://user-images.githubusercontent.com/580843/28471460-172f0f5a-6e35-11e7-88aa-ec8c236a78c0.png)

Fancy with 0.5 pixel shift:
![x-shift](https://user-images.githubusercontent.com/580843/28471483-240aef32-6e35-11e7-8db6-1e3dbf2e5403.png)

Fancy with shift and round up:
![x-shift-roundup](https://user-images.githubusercontent.com/580843/28471506-4207b25e-6e35-11e7-8aec-c31a2b7f5759.png)

My feeling is that the middle one is probably the best compromise. It'll work in all clients, it's slightly sharper than the old 2x2 code, and there's no top/left clipping.

Let's make the middle one the default and remove the two options.

**Comment by beaudet** [comment]
I definitely like the improvement in quality, but I'm still not understanding why, with Lanczos, there would still be smashing of the image up against the right side and the bottom?  See the flattening of the circle in those places?  I don't think this has anything to do with rounding up or down.  Proper distribution of pixel data should work at any resolution right?  That is, assuming the crop is circular to begin with, I would expect the edges to appear nearly identically and the circle retained almost flawlessly except at very low resolutions.

**Comment by jcupitt** [comment]
The problem is that the image is being reduced by x2 each time, but the canvas is not, since the original image is not a power of two in size. So you get very slight clipping since (with round down) the canvas can be up to 0.99 pixels smaller than it should be.

**Comment by beaudet** [comment]
Right, but even if you went from 1001 to 500 a resize operation that's assigning values from the source pixel to the target pixel proportionally to that geometry should not result in any clipping at all.  Are 2x reduction optimizations still being used even with the Lanczos filter in place that are leading to the clipping?  If so, I'd recommend moving away from the box filter entirely to focus on quality rather than performance with the fancy option.

**Comment by jcupitt** [comment]
A 1001 pixel across image with an exact x2 downsize would become 500.5 pixels across. If you round the width of the canvas down to 500, you will cut a bit off the right and bottom :-( which is what we are seeing here.

The solution is to either round the canvas size up, which will break some viewers, or to add a small shift so that all edges are clipped equally. My feeling is that adding the shift is probably the best compromise (that's the middle option above).

**Comment by beaudet** [comment]
Well obviously, the number of pixels in the target needs to be a whole number, but it doesn't really matter if you choose 500 or 501.  The resizing algorithms should allocate one or more pixels from the source image to pixels in the target image using weights - a Lanczos implementation should already be doing that no?  For example if I have the following nine pixel grayscale image and I'm reducing it to 4 pixels wide

ABC
DEF
GHI

to

JK
LM

then when calculating the pixel value for J in the target, you take the center of pixel J which is 0.5,0.5 and also calculate the center in the sample (0.5 * 3/2, 0.5 * 3/2) = (0.75, 0.75).  Then, in the case of a nearest neighbor weighted resample, I believe you would calculate the value of J as follows:

J = ( A/1 + B/2 + D/2 + E/4 ) / 2.25

2.25 is used since we're using 2.25 pixels in the source sample to calculate the value of one pixel in the target.

So, in the case of a 100% white image, let's say A = 255, B = 255, D = 255, E = 255, that would give:

J = ( 255 + 255 + 255 / 4 ) / 2.25 =  255

and no pixel data is lost in the resampling.  This holds true over the entire image, so no data is lost. 

The image below shows what I mean.  The dashed lines on the left represent the target image mapped onto the source and why the factors above were used to distribute the pixels from the source to the target.

![20170722_162059](https://user-images.githubusercontent.com/77699/28494359-f8877f04-6ef9-11e7-9d0e-01c7c054bb6c.jpg)

**Comment by jcupitt** [comment]
Yes, that's what it's doing, though the lanczos3 version takes a 12 x 12 pixel stencil, rather than 2 x 2.

The difficulty is the edges. A 3 x 3 pixel source would become 1.5 x 1.5 pixels with a x2 subsample. To make it concrete, suppose the image is all 255, and the surround is all 0. This source image:

```
255 255 255 0
255 255 255 0
255 255 255 0
0   0   0   0
```

Might become:

```
255 128 0
128 64  0
0   0   0
```

With round-down, the output image is 1 x 1, with round-up it's 2 x 2. With round-down you are effectively throwing away the right and bottom edges of the input, with round-up you have half pixels.

Round-up breaks some readers, so I was proposing a compromise with round-down plus a 0.5 pixel shift down and right (or up and left on the input) to make the clipping symmetrical. All edges get clipped by a hopefully almost invisible 0.25 pixels, rather than just the bottom and right by 0.5.

**Comment by beaudet** [comment]
In your example you generate a 3x3 before rounding down. If you round down first, you'd be targeting a 2x2 rather than a 3x3 and your result would instead be something closer to:

255 128
128 128

**Comment by beaudet** [comment]
Did my comment make sense or am I missing something basic?  What seems completely obvious is that software like Photoshop is able to resize images from 1000 to 501 or 500 just fine without any perceptible loss of pixel data.  It is my hope that this can also be done with vips.  I'm quite sure that think the Lanczos resize algorithm that's committed to IIP under a pull request achieves this as well.

**Comment by jcupitt** [comment]
Sure, and vips can too, but this is doing a fixed x2 shrink, not a variable shrink. Pyramids need exact x2 shrinks, you can't adjust them. 

Since you can't change the shrink ratio, you need to adjust the canvas size rounding instead.

**Comment by beaudet** [comment]
I understand that pyramids need exact (and arguably round down) x2 reduction.  I guess what I don't understand is why data would be lost when resizing from 1001 to 500 but not from 1001 to 501.  If the resize algorithm is solid, it shouldn't matter.  Are you saying the current algorithm relies on x2 optimizations and would have to be rewritten to be variable?  If so, I guess that's what I'm requesting.

**Comment by jcupitt** [comment]
Imagine a 1D case and exact x2 shrink with binning, so each output pixel is the simple average of two input pixels.

If the input is four pixels, like this:

```
A B C D
```

You then calculate:

```
A' = (A + B) / 2
B' = (C + D) / 2
```

to make the shrunk image:

```
A' B'
```

Lanczos3 is a fancier way of doing the averaging over a larger window, but the pixel geometry is the same.

Now suppose the input is 3 pixels across, `A B C`. You add a new pixel, `D`, at the end by copying `C` over one, then shrink. 

You now have `A' B`` as before, but `B'` is really a half-pixel. If you keep it, it's round up, if you throw it away, it's round down.

If you do simple round down, you'll find some images, like your roundel, will be trimmed slightly bottom and right. So the proposed solution adds a slight shift to make the clipping even on all sides.

**Comment by beaudet** [comment]
Why do you need to create a new column in the case of 3 pixels?  It seems like to reduce by x2 in the case of width == 3, you just divide 3 by 2 which, when rounded down, gives 1 pixel width.  So A' = avg(A, B, C), no?

And if you wanted to round up, you would be going from width = 3 (A, B, C) to width = 2 (A', B'), wouldn't it be:
A' = ( 1.0 * A + 0.5 * B ) / 1.5
and 
B' = ( 0.5 * B + 1.0 * C ) / 1.5

?

**Comment by jcupitt** [comment]
It's always two input pixels to one output pixel, no overlaps, so for 3 -> 1 you'd do 

```
A' = (A + B) / 2
```

and for 3 -> 2 it'd be

```
A' = (A + B) / 2
B' = (C + D) / 2
```

where D is the invented pixel.

**Comment by beaudet** [comment]
But that's not how all resizing algorithms work.  That's just how it's implemented in VIPS right?  Any reason it couldn't be changed to overlap?

**Comment by jcupitt** [comment]
It's how all x2 downsample algorithms work -- that's what x2 downsample means. If you want to allow overlaps, you're not shrinking by two, it's some other factor close to two, but not two.

**Comment by beaudet** [comment]
Seems like 1001 -> 500 (round down) would not be exactly x2 either, so regardless of whether it's x2 or something very close to x2 like x2.002, I think the most important characteristic is to ensure that all pixels in the source are mapped exactly to the target to retain the integrity of the original image.  Adding or subtracting columns to force fit an x2 optimization produces artifacts that will prevent me from using PTIF as the source for dynamically generated images of arbitrary size.  Would it be possible to use a variable resizing algorithm instead?

**Comment by jcupitt** [comment]
Oh dear, it feels like we're talking at cross purposes :( I wonder if the confusion is that there are two downsizing ratios here,

First, there's pixel mapping, ie. how many input pixels make one output pixel. In my opinion, this needs to be locked at exactly 2. This makes mapping pixel coordinates between layers simple, and it's the ratio that all viewers assume. It also makes calculation much, much simpler and faster. 

Second, there's the ratio of image sizes, or canvases. These have to be integers, so they can't always be locked at 2, unless the original source image is a power of 2, of course. 

Maybe consider a point _(x, y)_ in the shrunk image. What's the formula for finding the corresponding coordinate in the layer up? I think it should be simply _(2x, 2y)_.

**Comment by ruven** [comment]
Isn't the best solution, therefore, to round up the canvas size? Also, if you are clipping at resolution n, then your derivative image at resolution n+1 is going to be produced from an image that has already been clipped.

In terms of breaking viewers, what I can do for iipsrv is allow the user to define how rounding occurs server-side and have any clipping performed on the fly if required.

**Comment by beaudet** [comment]
If the intention is to retain the x2 optimization then yes, I agree that rounding up when necessary to ensure exactly x2 is the only way to avoid mangling the image at the edges.  From John's latest samples above, the round-up is the only option that seems to properly retain the circular crop.  If round-up is the winner, I'll take a closer look with a few additional circular crops and compare them to photoshop / ImageMagick.

**Comment by jcupitt** [comment]
It's keeping a large border (up to 6 pixels) around the image for the interpolation mask, it's just the images written to the output which are clipped, the derivatives should be OK.

Yes, we could round up, but I was worried about breaking viewers :( 

Maybe we should have a flag meaning round-up-plus-lanczos3 which turns on both of these things, and leave the default behaviour alone. At least people would have to opt into the possible breaking behaviour then.

Any suggestions for the flag name? I've become almost fond of `--fancy`.

**Comment by beaudet** [comment]
I revised my comment above.

**Comment by beaudet** [comment]
I definitely think you should leave the default behavior as-is to avoid breaking existing processes.

**Comment by beaudet** [comment]
John, what do you think about this?  Seems like we need a "--fancier" option that abandons the x2 optimized approach in favor of a more typical resizing algorithm that allocates sub-pixels from the source to the target.

**Comment by beaudet** [comment]
ping... hoping to come to some conclusion on this, one way or the other.

**Comment by jcupitt** [comment]
Hi again, sorry, I've been on other projects.

I really don't like the idea of varying the shrink ratio between layers. It would be a lot of work to implement and extremely slow, for what seems to me to be a marginal improvement. 

Are you sure shift + round up is unacceptable? 

![x-shift-roundup](https://user-images.githubusercontent.com/580843/28471506-4207b25e-6e35-11e7-8aec-c31a2b7f5759.png)

If @ruven is OK adding a thing to trim layers down again for viewers which assume round-down, it would work, I think.

**Comment by jcupitt** [comment]
Ooops, wrong button, sorry.

**Comment by ruven** [comment]
Rounding up is fine for me. Trimming is in fact already implemented in iipsrv for JPEG2000, as Kakadu implements rounding up for resolution levels.

**Comment by beaudet** [comment]
I'm afraid for these particular circular crops - we happen to have two in our highlights, I find the results to be noticeably different and my "eliminating all intermediate derivatives" strategy depends on picture perfect resizing, so it's pretty important to my particular use case.  Interesting point from Ruven about Kakadu.  Fortunately we don't use JPEG2000 sources and have no current plans to do so.  Would you recommend that I give ImageMagick another look rather than vips for the time being?  I'd obviously prefer to keep using vips and had previously discovered some corruption problems with ImageMagick's PTIF generation a few years back, but hopefully those have been resolved by now.  Aside from some potentially time consuming changes to vips, the other option would be to adjust the derivative generation process on my end to ensure the resolution is evenly divisible by 2 down to the desired tile size.  I would just have to sacrifice a few pixels for that I guess.  I can't think of any reason why that would be a horrible idea outside the context of registered images... but the registered images we produce are usually the exact same size anyway... as opposed to being overlaid on top with an offset.

**Comment by beaudet** [comment]
John, I've investigated ImageMagick as an alternative.  It seems to generate the tif pages fine now (no more corruption) and eliminates the slight color darkening that I've seen with vips although as you have pointed out on at least one post , it doesn't do chroma sub-sampling quite right unless the colorspace is ycbcr.  As a result the images it produces can be 3x larger than with vips.  Kind of a bummer.  Fortunately, IIP doesn't seem to be affected much by that increase so as best I can tell, it will only increase our storage requirements which I'm willing to accept for now.  However, if I have time in the next six months, I'd like to submit a pull request for vips that would offer an alternative to the box filter in order to avoid discarding pixels during the downsampling.  What do you think?

**Comment by jcupitt** [comment]
A quick comment after rereading this issue: 

The libvips tiff pyramid builder really has to use a strict /2 downsize ratio for pixels. If you want a downsize ratio that varies semi-randomly between pyramid layers, you would need to completely rewrite the pyramid code, and the result would be much slower and use a lot more memory. 

Probably the best way to do this would be as a separate program in Python or whatever that generated each layer in series and combined them at the end with `tiffcp`.

**Comment by beaudet** [comment]
What if the ratio continued to be strict in terms of dimensions from one layer to the next, i.e. truncate with integer arithmetic, but a Lanczos resampling algorithm was added as an additional option for translating down from the next higher layer?  It would be a bit slower yes, but it would avoid the problem of discarding pixels since all pixels from layer N+1 would be considered in the downsize. to layer N.

**Comment by jcupitt** [comment]
Hi again @beaudet, yes, that's what the `--fancy` option does, if I understand your meaning.

**Comment by jcupitt** [comment]
Oh I think you're asking for the image pixel downsize ratio to vary between layers. No, this is not possible due to the way that the libvips pyramid builder works. It must always be exactly two. You would need to rewrite the whole thing to change this.

You can have different downsize algorithms (NN, box, lanczos etc.), which is what the fancy branch is doing. You can also add an offset or change the way that the canvas size is rounded.

**Comment by beaudet** [comment]
From reading the history above to refresh my memory, I believe the problem with --fancy as implemented is that it's still not using a variable shrink ratio.  In other words, when starting out with a 1001x1001, it's not using all 1001 pixels in both dimension when computing the 500x500 image.  Instead, it's discarding the last column and row of pixels before applying Lanczos so the bounds of the original image actually shrink as layer computations continue.  I understand that the box downsize needs an even number of pixels and that odd ones need to be discarded, but Lanczos shouldn't have that constraint imposed in my opinion.

**Comment by beaudet** [comment]
Yup, that's what I'm asking.  I realize it's a lot of work.  I'll explore the tiffcp option though - thanks for that.  As it turns out, ImageMagick produced pyramids are about 3x as large as the ones vips produces.  It turns out that some library used by I.M. hasn't properly implemented chroma subsampling.  JPEGs produced outside of a tiff with I.M. don't seem to have this problem so it could be libtiff.

**Comment by jcupitt** [comment]
It doesn't really discard pixels on the input (the 1001 pixel image), it just doesn't generate the 501-st pixel in the output -- that is, it rounds the canvas size down by default. This makes a difference for lanczos3, which uses a 12x12 stencil rather than 2x2.

The `--round-up` flag makes it round up instead, so it always generates the extra column in the output for odd input.

**Comment by beaudet** [comment]
I recall seeing some minor problems even with -fancy + round-up, but it might be time to do some more tests with it including testing of the client side to ascertain support for round-up.  I know the IIP stack has been fixed to support this at least.

**Comment by beaudet** [comment]
I might have answered my own question: seems -t is probably required as tiffinfo reports Tile Width and Tile Length when using the -t option.  I'll have to research whether the embedded color profiles are retained as well.  If so, this will probably work well for me in which case I'll return here and document the exact commands I end up using.

**Comment by jcupitt** [comment]
Yes, that looks reasonable. thank you for uploading your script, David!

I thought of a few things:

You are shrinking down to 64x64, but your final tiles are 256x256. Should those be the same?

You are removing the profile, but then reinjecting an sRGB profile. Wouldn't no profile default to sRGB on most platforms? Perhaps you don't need to reinject.

You could make your script a little more automatic with:

```bash
input=$1
W=$(vipsheader -f width $input)
H=$(vipsheader -f height $input)
```

And it'll work for any input file type, not just tiff.

**Comment by jcupitt** [comment]
No, no one's made a Perl binding, as far as I know.

**Comment by beaudet** [comment]
Maybe one of these days, I'll write a Perl binding, but for now, the command line approach seems to be working very well.  I've revised the script which is pasted below.  Benchmarking IM against vips+tiffcp using our largest image (26700 x 26700 pixels) shows that the vips + tiffcp approach is six times faster (40 vs. 250 seconds) than ImageMagick, uses less than 30% of the memory (4GB vs. 14GB - and that's tiffcp - vips uses < 1GB even for the large image), and doesn't even saturate the processor cores like IM does in all cases (maybe tiffcp could be even faster?).  Furthermore, this approach produces a file that's nearly 3x smaller than with I.M. due to chroma subsampling support for jpeg compressed tiff files.  I'm getting similar results against a smaller image (11038x11038): 9 sec vs. 41 sec, < 1GB RAM vs. 3.2 GB, barely a blip on the CPU vs. CPU saturation, and a file that's only 35% the size of the one produced with ImageMagick.

Here's the improved script:

```#!/bin/bash

function handle_error() {
  echo "Script exited with status ${2} at line ${1}"
}

trap 'handle_error ${LINENO} $?' ERR

# cause the script to fail if any commands exist with non-zero status
set -e

if [[ -z $1 || -z $2 || -z $3 ]]; then
    echo "Usage:make_pyramidal_tiff.bash <tmpdir> <full path to image source> <full path to output target>";
    exit 1;
fi
if [[ ! -d $1 || ! -f $2 ]]; then
    echo "Either the tmpdir or input file do not exist.";
    exit 1;
fi

if [ -f $3 ]; then
    echo "The output file already exists! Aborting.";
    exit 1;
fi

touch $3;

if [ ! -f $3 ]; then
    echo "Could not create the output file.";
fi

\rm $3

tmpdir=$1
input=$2
outfile=$3
pid=$$;
tmpprefix=${tmpdir}/stripped_tiff_${pid}
outprefix=${tmpdir}/icc_tiff_${pid}

# cleanup temp files that might already exist
\rm $tmpprefix* || true
\rm $outprefix* || true

W=$(vipsheader -f width $input)
H=$(vipsheader -f height $input)
c=0;
while [ 1 ]; do
    # echo ${c}, ${W}, ${H}
    if (( c > 0 )); then
        d=$((c - 1))
        # since we already have a stripped and color converted tiff that
        # is twice the resolution as the one are about to create, use that
        # one instead of the original when resizing which will save quite
        # a bit of processing time
        vipsthumbnail ${outprefix}_$d.tif --size $Wx$H -o ${outprefix}_$c.tif[compression=none]
    else
        # create a native resolution uncompressed tiff with all metadata stripped out to save space
        vipsthumbnail ${input} --size $Wx$H -o ${tmpprefix}_$c.tif[compression=none,strip] --eprofile /usr/local/nga/etc/sRGBProfile.icc
        # then re-embed the ICC profile since we want to have an ICC in the file pyramid
        vipsthumbnail ${tmpprefix}_$c.tif --size $Wx$H -o ${outprefix}_$c.tif[compression=none,profile=/usr/local/nga/etc/sRGBProfile.icc]
    fi

    # reduce height and width by half and repeat process until small enough
    W=$(( W / 2));
    H=$(( H / 2));
    c=$(( c + 1));
    if (( W < 1 || H < 1 || (( W < 129 && H < 129 )) )); then
        break;
    fi
done

# once we have all sizes created, then we use tiffcp to perform the pyramid
# assembly and jpeg compression at 90 quality

# note: tiffcp defaults to ycbcr photometric interpretation and presumably therefore
# chroma subsampling so by default it produces JPEGs about 3x smaller with ycbcr
# vs. when photometric interpretation is set to rgb
# e.g. -c jpeg:r:90 vs. -c jpeg:90
tiffcp -c jpeg:90 -t -w 256 -l 256 ${outprefix}_*.tif -a ${outfile}

# cleanup temp files but leave the outputput file in place
\rm $tmpprefix*
\rm $outprefix*

# sanity check
WF=$(vipsheader -f width $outfile)
HF=$(vipsheader -f height $outfile)

echo "Pyramid width: ${WF}"
echo "Pyramid height: ${HF}"

# final check
exit $(( WF == W && HF == H ))

```

**Comment by beaudet** [comment]
two problems I ran into:
1. there's a bug in the script $Wx$H needs to be changed to ${W}x${H}
2. vipsthumbnail doesn't honor pixel-perfect resize requests - presumably it's deciding to round up or down based on the original aspect ratio

# vipsheader icc_tiff_18403_0.tif
icc_tiff_18403_0.tif: 15918x10823 uchar, 3 bands, srgb, tiffload

# vipsthumbnail icc_tiff_18403_0.tif --size 7959x5411 -o icc_tiff_18403_1.tif[compression=none]

# vipsheader icc_tiff_18403_1.tif
icc_tiff_18403_1.tif: 7958x5411 uchar, 3 bands, srgb, tiffload

I requested 7959x5411 but vipsthumbnail returned 7958x5411.  I assume if I switch to using the standard vips command line, I can work around this.  Time to try that out.

**Comment by jcupitt** [comment]
It doesn't try to calculate the horizontal and vertical scale factors together. Instead, it does one then the other and picks the smaller.

In your example, the calculation is:

horizontal scale factor = 7959 / 15918
    = 0.5
vertical scale factor = 5411 / 10823
    = 0.49995

So it picks the vertical scale. Then the new horizontal size is:

new horizontal size = 15918 * 5411 / 10823
    = 7958.2646216391

Then round to nearest gives 7958.

You can force it to break the aspect ratio with a `"!"` in the size specification:

```
vipsthumbnail icc_tiff_18403_0.tif --size 7959x5411! -o icc_tiff_18403_1.tif[compression=none]
```

It might be better in your case.

**Comment by jcupitt** [comment]
You're right, there was no note about `!` in the docs, I added something. Thanks!

**Comment by beaudet** [comment]
Sure thing.  I ended up using a combination of vips icc_transform and vips tiffsave to create the initial color transformed image with an injected ICC profile, then vipsthumbnail with the WxH! syntax for everything else.  The updated script is below and is running on a test of about 1200 images.  It will be interesting to see how much faster this runs than the ImageMagick version since I'm able to ramp up the parallel processes to 8 from 2 due to the much lower CPU and memory constraints. 

```
#!/bin/bash

function handle_error() {
  echo "Script exited with status ${2} at line ${1}"
}

trap 'handle_error ${LINENO} $?' ERR

# cause the script to fail if any commands exist with non-zero status
set -e

if [[ -z $1 || -z $2 || -z $3 ]]; then
    echo "Usage:tiff_to_pyramid.bash <tmpdir> <full path to image source> <full path to output target>";
    exit 1;
fi
if [[ ! -d $1 || ! -f $2 ]]; then
    echo "Either the tmpdir or input file do not exist.";
    exit 1;
fi

if [ -f $3 ]; then
    echo "The output file already exists! Aborting.";
    exit 1;
fi

touch $3;

if [ ! -f $3 ]; then
    echo "Could not create the output file.";
fi

\rm $3

tmpdir=$1
input=$2
outfile=$3
pid=$$;
tmpprefix=${tmpdir}/stripped_srgb_${pid}
outprefix=${tmpdir}/srgb_${pid}

# cleanup temp files that might already exist
\rm $tmpprefix* || true
\rm $outprefix* || true

# first, run an icc_transform to convert the original to sRGB and strip all metadata from the file
vips icc_transform $input ${tmpprefix}.tif[compression=none,strip] /usr/local/nga/etc/sRGBProfile.icc
# now, embed the ICC profile since it is stripped out with the strip metadata directive
vips tiffsave ${tmpprefix}.tif ${outprefix}_0.tif --compression none --profile /usr/local/nga/etc/sRGBProfile.icc

# read the width and height of the transformed file
W=$(vipsheader -f width ${outprefix}_0.tif)
H=$(vipsheader -f height ${outprefix}_0.tif)
c=0;
while [ 1 ]; do
    W=$(( W / 2 ));
    H=$(( H / 2 ));
    #echo ${c} ${W} ${H}

    # since we already have a stripped and color transformed tiff that
    # is twice the resolution as the one are about to create, use that
    # one instead of the original when resizing which will save quite
    # a bit of processing time
    vipsthumbnail ${outprefix}_$c.tif --size ${W}x${H}\! -o ${outprefix}_$(( c + 1 )).tif[compression=none]

    # reduce height and width by half and repeat process until small enough
    if (( W < 1 || H < 1 || (( W < 129 && H < 129 )) )); then
        break;
    fi
    c=$(( c + 1 ));
done

# once we have all sizes created, then we use tiffcp to perform the pyramid
# assembly and jpeg compression at 90 quality

# note: tiffcp defaults to ycbcr photometric interpretation and presumably therefore
# chroma subsampling so by default it produces JPEGs about 3x smaller with ycbcr
# vs. when photometric interpretation is set to rgb
# e.g. -c jpeg:r:90 vs. -c jpeg:90
tiffcp -c jpeg:90 -t -w 256 -l 256 ${outprefix}_*.tif -a ${outfile}

# cleanup temp files but leave the outputput file in place
\rm $tmpprefix*
\rm $outprefix*

# sanity check
WF=$(vipsheader -f width $outfile)
HF=$(vipsheader -f height $outfile)

echo "Pyramid width: ${WF}"
echo "Pyramid height: ${HF}"

# final check
exit $(( WF == W && HF == H ))
```

**Comment by beaudet** [comment]
I re-ran a test of 1147 images ( 200 GB of source images with average resolution of 7000 x 7200 ) including downloading the original assets from Amazon S3 prior to processing and it completed without any errors in 85 minutes.  That's quite a lot faster than with ImageMagick.   I was also able to run 24 simultaneous vips processes which seemed to do a reasonably good job of maximizing the 8 CPU cores without coming close to exhausting the server's 48GB of RAM).  I can almost certainly cut the RAM by 50% and repeat without any problems.

**Comment by scossu** [comment]
I have followed this thread with interest and I am converting @beaudet 's script into Python. Ideally I would like to use only pyvips. 

Basically what I am planning to do is creating a number of images with decreasing size using an arbitrary kernel (say Lanczos3), then assembling them in a multi-page TIFF file. 

I noticed the `pyramid` flag of [tiffsave](https://libvips.github.io/pyvips/vimage.html#pyvips.Image.tiffsave) that automates the derivative generation. So I created a test square image with a diamond whose corners touch the edges of the image and used this script on it: 

    im1 = pyvips.Image.new_from_file('/home/scossu/tmp/test_diamond_3001.tif') 
    im1.tiffsave(
        '/home/scossu/tmp/pyramid_diamond_3001_pyvips.tif',
        tile=True, tile_width=256, tiile_height=256, pyramid=True
    )

I notice no artifacts. I tried with a 4096x4096 image and a 3001x3001 image derived from the former. See screenshot from the 3001 image opened in Gimp, showing the stacked pyramid images:

![2019-02-23_845x652](https://user-images.githubusercontent.com/4306733/53296522-760a9200-37c5-11e9-8a7c-e64aa2c0dc03.png)

Am I supposed to see similar issues as @beaudet 's round image?

**Comment by jcupitt** [comment]
Hello @scossu,

Yes, the tiff saver has a very fancy pyramid builder built in. It's able to write all the layers of the pyramid in parallel in a single pass through the image, so it's fast and only uses a little memory. 

It has a couple of issues though:

1. Because of the way it works, it can only do a exact x2 shrink for each layer. @beaudet needed to vary both the shrink factor and the aspect ratio between layers to get his circular images to fit exactly. 
2. It does support different types of downsampling, but only 2x2 mean, mode and median. Lanczos is very slightly sharper, though it doesn't feel like a huge difference to me. Perhaps I need better glasses.

There's a branch up there ^^ which adds lanczos, though it was never merged.

I think if you make a perfect circle that exactly touches the image edges and then pyramid it, you may notice slight flattening of the right and bottom edges in some layers.

Thinking about this again, simply adding one pixel of white along the bottom and right edge of the input would be enough to prevent that. For example:

```
W=$(vipsheader -f width $input)
H=$(vipsheader -f height $input)
vips gravity A15.tif my-pyr.tif[pyramid,tile,compression=jpeg,Q=80] north-west $((W+1)) $((H+1)) --extend white
```

To create `my-pyr.tif`.

**Comment by jcupitt** [comment]
We should close this -- let's start a new issue if there's still a problem here.

**Comment by scossu** [comment]
Thanks @jcupitt. A couple of last post-closing related questions: 

>   2. It does support different types of downsampling, but only 2x2 mean, mode and median. Lanczos is very slightly sharper, though it doesn't feel like a huge difference to me. Perhaps I need better glasses.

Is there a parameter to  specify the down-sampling algorithm in the python function? (I don't see enums for existing algorithms in the Python API and I'm not sure what mean, mode and median correspond to.)

Also, for full customization to replicate @beaudet 's specific tweaks, I though I could combine [resize](https://libvips.github.io/pyvips/vimage.html#pyvips.Image.resize) and [insert](https://libvips.github.io/pyvips/vimage.html#pyvips.Image.insert), but it seems like `insert` flattens the image. Is there a straightforward way to insert an image as a new layer?

**Comment by jcupitt** [comment]
It's like any optional argument. Try:

```
$ vips tiffsave
save image to tiff file
usage:
   tiffsave in filename [--option-name option-value ...]
where:
   in           - Image to save, input VipsImage
   filename     - Filename to save to, input gchararray
optional arguments:
... etc.
   region-shrink - Method to shrink regions, input VipsRegionShrink
			default: mean
			allowed: mean, median, mode
```

So in py you can write:

```python
image = pyvips.Image.new_from_file("input/filename")
image.write_to_file("output/filename.tif", region_shrink="mode")
```

or

```python
image.tiffsave("somefilename", region_shrink="median")
```

Mean, mode and median are how it shrinks each 2x2 block of pixels. For

```
a b 
c d
```

Mean is just `(a + b + c + d) / 4`, median is `[a, b, c, d].sort()[1]`, mode is the most frequent value. Median and mode are useful for shrinking images where pixel values are not lightness, but something like label or index.

Docs here:

https://libvips.github.io/libvips/API/current/VipsForeignSave.html#vips-tiffsave

To implement the pyramiding yourself (I would only do this is you really MUST have lanczos or sub-pixel perfect alignment of edges), you'll need to make a set of separate TIFF images and combine them with tiffcp. libvips does support writing many-page TIFFs, but all the pages have to be the same size, so it won't work for pyramids.

**Comment by scossu** [comment]
Ah. It was the `shrink_region` parameter that I did not quite grasp. It looks like the default value is good enough for now, anyways. Thanks again.

**Comment by beaudet** [comment]
For anyone looking for the latest tiff_to_pyramid.bash script, you can find it here: https://github.com/NationalGalleryOfArt/iipsrv/blob/master/imagescripts

@scossu has also turned this script into Python code
